TorchServe IPEX Blog #2

min-jean-cho opened the pull request.

msaroufim reviewed (approved):
Fantastic blog (as usual) - thank you @min-jean-cho

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+Graph Optimization
+------------------
+It is highly recommended for users to take advantage of Intel® Extension for PyTorch* with `TorchScript <https://pytorch.org/docs/stable/jit.html>`_ for further graph optimizations. To optimize performance further with TorchScript, Intel® Extension for PyTorch* supports oneDNN fusion of frequently used FP32/BF16 operator patterns, like Conv2D+ReLU, Linear+ReLU, etc to reduce operator/kernel invoc ...
```
```suggestion
It is highly recommended for users to take advantage of Intel® Extension for PyTorch* with `TorchScript <https://pytorch.org/docs/stable/jit.html>`_ for further graph optimizations. To optimize performance further with TorchScript, Intel® Extension for PyTorch* supports oneDNN fusion of frequently used FP32/BF16 operator patterns, like Conv2D+ReLU, Linear+ReLU, or other, to reduce operator/kernel invocation overheads, and for better cache locality. Some operator fusions allow to maintain temporary calculations, data type conversions, data layouts for better cache locality. As well as for INT8, Intel® Extension for PyTorch* has built-in quantization recipes to deliver good statistical accuracy for popular DL workloads including CNN, NLP and recommendation models. The quantized model is then optimized with oneDNN fusion support. 
```

min-jean-cho replied:
have changed etc to and more.

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+^^^^^^^^ 
+Let's profile FP32 graph optimization with TorchScript. 
+
+As in the previous exercises, we'll bind the workload to physical cores of the first socket.
```
```suggestion
As in the previous exercises, we will bind the workload to physical cores of the first socket.
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+---------------------------
+When invoking *ipex.optimize* on model, Intel® Extension for PyTorch* automatically converts the model to optimized memory format, channels last. Channels last is a memory format that is more friendly to Intel Architecture. Compared to PyTorch default channels first NCHW (batch, channels, height, width) memory format, channels last NHWC (batch, height, width, channels) memory format generally acc ...
+
+One thing to note is that it is expensive to convert memory format. So it's better to convert the memory format prior to deployment once, and keep the memory format conversion minimum during deployment. As the data propagates through model's layers the channels last memory format is preserved through consecutive channels last supported layers (e.g., Conv2d -> ReLU -> Conv2d) and conversions are o ...
```
```suggestion
One thing to note is that it is expensive to convert memory format. So it's better to convert the memory format prior to deployment once, and keep the memory format conversion minimum during deployment. As the data propagates through model's layers the channels last memory format is preserved through consecutive channels last supported layers (for example, Conv2d -> ReLU -> Conv2d) and conversions are only made in between channels last unsupported layers. See `Memory Format Propagation <https://www.intel.com/content/www/us/en/develop/documentation/onednn-developer-guide-and-reference/top/programming-model/memory-format-propagation.html>`_ for more details.
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+        model = torch.jit.trace(model, data)
+        model = torch.jit.freeze(model)
+
+We'll use `oneDNN verbose mode <https://oneapi-src.github.io/oneDNN/dev_guide_verbose.html>`_, a tool to help collect information at oneDNN graph level such as operator fusions, kernel execution time spent on executing oneDNN primitives. For more information, refer to the `oneDNN Documentation <https://oneapi-src.github.io/oneDNN/index.html>`_.
```
```suggestion
We will use `oneDNN verbose mode <https://oneapi-src.github.io/oneDNN/dev_guide_verbose.html>`_, a tool to help collect information at oneDNN graph level such as operator fusions, kernel execution time spent on executing oneDNN primitives. For more information, refer to the `oneDNN Documentation <https://oneapi-src.github.io/oneDNN/index.html>`_.
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+========================================================= 
+Intel® Extension for PyTorch* has already been integrated into TorchServe to improve the performance out-of-box. :superscript:`2` When TorchServe creates single or multiple instances for inference, the launcher from Intel® Extension for PyTorch* is invoked automatically to set CPU affinity to these instances when launcher is enabled. :superscript:`3` This helps to make each instance use its assig ...
+
+2. The feature has to be explicitly enabled by setting *ipex_enable=true* in  *config.properties*.
```
should lines 383 and 384 be converted int one?

min-jean-cho replied:
Let's keep the two features (and lines 383, 384) separate to be explicit.

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+========================
+Let's profile Intel® Extension for PyTorch* optimizations with TorchServe. 
+
+We'll use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32 TorchScript, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 
```
```suggestion
We will use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32 TorchScript, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+We'll use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32 TorchScript, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 
+
+As in the previous exercise, we'll use the launcher to bind the workload to physical cores of the first socket. To do so, user simply needs to add a few lines in `config.properties <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_: 
```
```suggestion
As in the previous exercise, we will use the launcher to bind the workload to physical cores of the first socket. To do so, user simply needs to add a few lines in `config.properties <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_: 
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+   :width: 100%
+   :align: center
+   
+Level-1 TMA shows that both are bounded by the backend. As discussed earlier, the majority of un-tuned deep learning workloads will be Back End Bound. Notice the Back End Bound reduced from 70.0 to 54.1. Let's go one level deeper. 
```
```suggestion
Level-1 TMA shows that both are bounded by the backend. As discussed earlier, the majority of untuned deep learning workloads will be Back End Bound. Notice the Back End Bound reduced from 70.0 to 54.1. Let's go one level deeper. 
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+The timeline graph can be expanded to see op-level profiling results. Notice that Conv + ReLU has been fused, and the duration decreased from 6.393 ms + 1.731 ms to 3.408 ms - 2.4x speedup. 
+
+**********
+Take-aways
```
```suggestion
Conclusion
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+**********
+Take-aways
+********** 
+In this blog, we've used Top-down Microarchitecture Analysis (TMA) and Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT) to demonstrate that 
```
```suggestion
In this tutorial, we have used Top-down Microarchitecture Analysis (TMA) and Intel® VTune™ Profiler's Instrumentation and Tracing Technology (ITT) to demonstrate that 
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+=====================================================================
```
```suggestion
```

svekars replied:
Can we please use the following heading syntax:
```
Title
===== 

Heading 1
-----------
Heading 2
~~~~~~~~~~
Heading 3
++++++++++

min-jean-cho replied:
Thanks @svekars, could you please tell me the syntax for Heading 4

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+   
+Throughout this blog, we'll use `Top-down Microarchitecture Analysis (TMA) <https://www.intel.com/content/www/us/en/develop/documentation/vtune-cookbook/top/methodologies/top-down-microarchitecture-analysis-method.html>`_ to profile and show that the Back End Bound (Memory Bound, Core Bound) is often the primary bottleneck for under-optimized or under-tuned deep learning workloads, and we'll demo ...
+
+*****************
```
We typically don't add a table of contents like this because it's autogenerated on the right hand side under Shortcuts. It is difficult to keep a manual TOC like this up-to-date with any future changes in the content.

min-jean-cho replied:
I'm fine with either options, but I thought TOC would help when readers read the intro since this tutorial has lots of content. Let me know what you'd prefer !

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+        - Intel® Extension for PyTorch* Optimizations  
+        - Intel® Extension for PyTorch* with TorchServe 
+        - Exercise
+        
```
Where is prerequisites? In the TOC above they are listed as first section.

min-jean-cho replied:
Do we need to have an explicit heading with "Prerequisites" ? The TOC is more of intended for a glance of content at one shot, rather than a link to the sections.

svekars replied:
It would be good to have a prerequisites section and list what the user need to know / have installed before they can complete this tutorial.

min-jean-cho replied:
OK. Let me try moving around some paragraphs, to heave content under each heading.

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+Intel® Extension for PyTorch*
+=============================
+`Intel® Extension for PyTorch* <https://github.com/intel/intel-extension-for-pytorch>`_ is a Python package to extend PyTorch with optimizations for extra performance boost on Intel hardware. Example optimizations use AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX).
```
```suggestion
`Intel® Extension for PyTorch* <https://github.com/intel/intel-extension-for-pytorch>`__ is a Python package to extend PyTorch with optimizations for extra performance boost on Intel hardware. Example optimizations use AVX-512 Vector Neural Network Instructions (AVX512 VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX).
```

svekars commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+In this section, we've demonstrated that JeMalloc can give better performance than the default PyTorch memory allocator, PTMalloc, with efficient thread-local caches improving Back End Bound.
+
+Intel® Extension for PyTorch*
```
This almost looks like a separate tutorial?

min-jean-cho replied:
It's part of this tutorial, a short introduction to Intel® Extension for PyTorch*, before jumping into TorchServe with Intel® Extension for PyTorch*

svekars reviewed (commented):
A few editorial suggestions. Also, need to fix these links:
```
/var/lib/jenkins/workspace/intermediate/torchserve_with_ipex.rst:2: WARNING: Duplicate explicit target name: "intel® vtune™ profiler".
 /var/lib/jenkins/workspace/intermediate/torchserve_with_ipex_2.rst:3: WARNING: Duplicate explicit target name: "config.properties"
```
Probably, just need to add double underscores in the link (`my link <url>`__)

svekars commented:
Also, I've noticed a few images are too wide which creates formatting issues:

![image1](https://user-images.githubusercontent.com/5317992/195459689-5cd6a6de-0278-4057-8cfc-5fb3cce58832.png)
![image3](https://user-images.githubusercontent.com/5317992/195459682-067edb4f-493c-4c6d-90a9-8710c34d400c.png)
![image2](https://user-images.githubusercontent.com/5317992/195459685-9ffd04c3-ae9c-4550-a8ed-1c437060b9cd.png)

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+Grokking PyTorch Intel CPU performance from first principles (Part 2)
+=====================================================================
+
+Authors: Min Jean Cho, Jing Xu, Mark Saroufim
```
Thanks @svekars for the review. Have done so.

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+Tune for the Back End Bound
+===========================
+The majority of un-tuned deep learning workloads will be Back End Bound. Resolving Back End bound is often resolving sources of latency causing retirement to take longer than necessary. As shown above, Back End Bound has two submetrics – Core Bound and Memory Bound. 
```
Let's keep as "Back End Bound" to be consistent with the flowchart.

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+===========================
+The majority of un-tuned deep learning workloads will be Back End Bound. Resolving Back End bound is often resolving sources of latency causing retirement to take longer than necessary. As shown above, Back End Bound has two submetrics – Core Bound and Memory Bound. 
+
+Memory Bound stalls have causes related to the memory subsystem. For example, last-level cache (LLC or L3 cache) miss causing access to DRAM. Scaling deep learning models often requires significant compute. And high compute utilization requires that data is available when the execution units need it to execute the uOps. This requires prefetching the data and reusing the data in cache instead of f ...
```
Let's keep as "significant compute".

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+*********************************************************************
+The ITT APIs of Intel® VTune Profiler is a useful tool to annotate a region of your workload for tracing to profile and visualize at a finer granularity of your annotation – OP/function/sub-function granularity. By annotating at the granularity of your PyTorch model's OPs, Intel® VTune Profiler's ITT enables op-level profiling. Intel® VTune Profiler's ITT has been integrated into `PyTorch Autogra ...
+
+1. The feature has to be explicitly enabled by with *torch.autograd.profiler.emit_itt()*.
```
Have changed to "enabled by `with torch.autograd.profiler.emit_itt()`"

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+
+1. The feature has to be explicitly enabled by with *torch.autograd.profiler.emit_itt()*.
+
+*********************************************
```
This is the intended TOC in detail (it's little bit hard to see this as a whole going through the rst file) :
![image](https://user-images.githubusercontent.com/93151422/195467746-407f277e-aa48-4421-bab7-31a4601fbaa4.png)

min-jean-cho replied:
I'm not sure if I understand fully, but they are not empty. `TorchServe with Intel® Extension for PyTorch*` is a big section with subsections `Leveraging Advanced Launcher Configuration: Memory Allocator`, `Intel Extension for PyTorch`, etc. Similarly, `Leveraging Advanced Launcher Configuration: Memory Allocator` is not empty, it's a section with subsection `TCMalloc, JeMalloc, PTMalloc`.

svekars replied:
Here is the suggestion on what to add to these sections so we don't have three headings one after another:

TorchServe with Intel® Extension for PyTorch*
----------------------------------------------
Intel® Extension for PyTorch*  is a PyTorch extension that provides performance acceleration for deep learning inference and training on Intel CPU. This section provides an example of how you can leverage this extension with TorchServe 

Leveraging Advanced Launcher Configuration: Memory Allocator
~~~~~~
Move line 72 here. 

TCMalloc, JeMalloc, PTMalloc
^^^^^
Both TCMalloc and JeMalloc use thread-local ....

min-jean-cho commented:
> Also, I've noticed a few images are too wide which creates formatting issues:

Thank you @svekars for checking on this. I've fixed the img sizes, could you check if they look fine now?

> We typically don't add a table of contents like this because it's autogenerated on the right hand side under Shortcuts.

Could I have a look at this expanded (maybe a screenshot)? I want to see if they show up as I've intended.

min-jean-cho commented:
> /var/lib/jenkins/workspace/intermediate/torchserve_with_ipex.rst:2: WARNING: Duplicate explicit target name: "intel® vtune™ profiler".
>  /var/lib/jenkins/workspace/intermediate/torchserve_with_ipex_2.rst:3: WARNING: Duplicate explicit target name: "config.properties"
> Probably, just need to add double underscores in the link (my link <url>__)

May I know the source of this warning if it's from an internal CI? I've tried fixing to my link <url>__ where applicable, but not sure if this will resolve that warning.

min-jean-cho commented on intermediate_source/torchserve_with_ipex_2.rst:
```diff
@@ -0,0 +1,471 @@
+Additionally, let's profile with PyTorch Profiler. 
+
+.. figure:: /_static/img/torchserve-ipex-images-2/13.png
+   :width: 150%
```
All images are already changed to less than 100% width. The comment is on outdated file, prior to this change.

min-jean-cho commented:
Thanks @svekars for the review, I have updated according to your comments. The updates are:
- Fixed syntax of title and headings
- Added content under each heading (moved around some existing paragraphs)
- Removed TOC 
- all img sizes are less than 100%

svekars reviewed (approved):
LGTM! Thank you for addressing the comments.

min-jean-cho commented:
Thanks @svekars for the thorough review - grammatical errors, also I realize it's easier to read now with content under each heading.  Thanks :) 

Will you be merging this?

min-jean-cho commented on index.rst:
```diff
@@ -496,6 +496,13 @@ What's new in PyTorch tutorials?
+   :card_description: A case study on the TorchServe inference framework optimized with Intel® Extension for PyTorch (Part 2).
+   :image: _static/img/thumbnails/cropped/generic-pytorch-logo.png
+   :link: intermediate/torchserve_with_ipex_2
+   :tags: Model-Optimization,Production
```
Hi @svekars, I see this tutorial has been published to pytorch.org [here](https://pytorch.org/tutorials/intermediate/torchserve_with_ipex_2.html?highlight=grokking%20pytorch%20intel%20cpu%20performance%20from%20first%20principles%20part). However, I don't see link to this tutorial in the left nav, could you pls. help check on this? 
![Screenshot (1374)](https://user-images.githubusercontent.com/93151422/196245046-b5cc89ae-3155-4ae2-a4f6-d7b81b01c903.png)
